xarray/core/dataset.py

```diff
@@ -4030,17 +4035,22 @@ def unstack(
         for dim in dims:
 
             if (
-                # Dask arrays don't support assignment by index, which the fast unstack
-                # function requires.
-                # https://github.com/pydata/xarray/pull/4746#issuecomment-753282125
-                any(is_duck_dask_array(v.data) for v in self.variables.values())
-                # Sparse doesn't currently support (though we could special-case
-                # it)
+                # Dask arrays supports assignment by index,
+                # https://github.com/dask/dask/pull/7393
+                (
+                    dask_version < "2021.04.0"
+                    and any(is_duck_dask_array(v.data) for v in self.variables.values())
+                )
+                # Sparse now supports kwargs in full_like,
                 # https://github.com/pydata/sparse/issues/422
-                or any(
-                    isinstance(v.data, sparse_array_type)
-                    for v in self.variables.values()
+                or (
+                    sparse_version < "0.11.2"
+                    and any(
+                        isinstance(v.data, sparse_array_type)
+                        for v in self.variables.values()
+                    )
                 )
+                # Shifting the arrays to sparse requires _unstack_full_reindex:
```

Comment:
This is good code! Though because of the L4054, this big `if` will still return `False`.

Does pint still break on if we uncomment that?

We can merge this because it gets us closer to this working (and it's generally better code than mine!), though it won't quite change behavior yet.

Reply:
I'm not sure I follow, `_unstack_once` is the fast path right? Which you get to when the if is `False`. So normal dask arrays for example should now be getting the fast path, which I hope will be a great thing!

I'm not completely sure what `sparse=True` means when already using all sparse arrays though?

Reply:
But then isn't this going to return True and the dask array is going to use `_unstack_full_reindex`? https://github.com/pydata/xarray/pull/5315/files#diff-763e3002fd954d544b05858d8d138b828b66b6a2a0ae3cd58d2040a652f14638R4064-R4066

Reply:
Ahh, I see now. It gets rather annoying having to special case all these optional imports. Can't we just remove this check and let it error instead? We could also add a parameter to force the slow path for the users that wants/has to use it.

Reply:
What's the case that we'd let it error? Currently sparse, pint & dask arrays all run on the slow path; none of them error.

Reply:
Can't always brainstorm good ideas. :) But switching around the if so the fast path is opt in seems to do the trick!